### ISSUE TITLE
Native support for nftables

### ISSUE THREAD

**Issue by rforberger** [docker/for-linux#1472]
Is there any configuration option to configure dockerd / moby / containerd to use nftables natively for firewalling?

I have a debian machine with docker and nftables, but my docker iptables rules get overwritten by nftables once they get restarted / reloaded.

My alternatives are set to iptables-nft for iptables as per Debian 12.

Does docker support nftables?

Ideally I would want my nftables rules be _dynamically_ merged with any docker firewall rules.

Thanks in advance.

**Comment by rforberger** [comment]
I don't have problems with nftables, but with docker's iptables implementation.

As from my understanding it is still using iptables right, as it creates iptables rules?

I don't want to migrate to legacy iptables-legacy backend.

If docker is already using nftables, why do my nftables the overwrite docker's iptables rules, when reloading?

**Comment by jeromecst** [comment]
> If docker is already using nftables, why do my nftables the overwrite docker's iptables rules, when reloading?

If your nftables.conf contains `flush ruleset`, it will flush all previous rules (including docker's) when reloading

**Comment by trallnag** [comment]
Is this really a problem with Docker? As @jeromecst mentioned, if the primary nftables.conf contains `flush ruleset`, which is often the case, dynamically added rules will be removed.

One solution would be for Docker to write it's rules to a dedicated file and include this file from the primary nftables.conf.

**Comment by rforberger** [comment]
No.

**Comment by wcasanova** [comment]
> As from my understanding it is still using iptables right, as it creates iptables rules?
> If docker is already using nftables, why do my nftables the overwrite docker's iptables rules, when reloading?

Actually docker does not use nftables natively, from debian 10 onwards nftables is the default framework, debian installs the iptables-nft package which translates iptables rules to nft rules.

**Comment by bluikko** [comment]
> > If docker is already using nftables, why do my nftables the overwrite docker's iptables rules, when reloading?
> 
> If your nftables.conf contains `flush ruleset`, it will flush all previous rules (including docker's) when reloading

Without `flush ruleset` all the rules will be duplicated since the new ruleset will just be added to the already existing chain.  
I guess that could be worked around with just `flush table` / `flush chain` instead.

Do you have this working? Wondering because the `iptables` tools stop working after modifications with `nft` on EL9 -- is Docker still able to change its rules after direct modifications to nftables?

I believe the point of this issue is that sooner or later iptables wrappers will be deprecated and nftables is the only way forward.

**Comment by rforberger** [comment]
I didn’t try any longer. I think docker should natively understand nftables rulesets, but still iptables. Otherwise in my opinion it’s useless if you have an existing nftables ruleset and want to load the rules from docker into it.Or did I completely misunderstand the concept?Sent from my iPhoneOn 3. Apr 2024, at 13:26, bluikko ***@***.***> wrote:﻿


If docker is already using nftables, why do my nftables the overwrite docker's iptables rules, when reloading?

If your nftables.conf contains flush ruleset, it will flush all previous rules (including docker's) when reloading

Without flush ruleset all the rules will be duplicated since the new ruleset will just be added to the already existing chain.
I guess that could be worked around with just flush table / flush chain instead.
Do you have this working? Wondering because the iptables tools stop working after modifications with nft on EL9 -- is Docker still able to change its rules after direct modifications to nftables?
I believe the point of this issue is that sooner or later iptables wrappers will be deprecated and nftables is the only way forward.

**Comment by ghost** [comment]
I agree with the author it's a mess to work like that on distro that use nftables as main way to manage firewall.

**Comment by cpuguy83** [comment]
nftables support is being worked on, however it is not a small change and cuts through just about every bit of the network stack.

**Comment by jeromecst** [comment]
In the meantime, `destroy` has recently been added on nftables. Consider using it inside your `nftables.conf` instead of `flush ruleset` to avoid flushing every docker rules upon reload.

https://git.netfilter.org/nftables/commit/?id=e1dfd5cc4c46514a84dd8a2063b45517b596e1ca

**Comment by jcmcken** [comment]
Just want to throw some comments in here. These are based on RHEL9 or compatible systems with a very simple network setup (single main interface ``eth0``).

Installing ``iptables-nft`` package and running ``alternatives --set iptables /usr/sbin/iptables-nft`` gets you almost all the way there. (This allows ``dockerd`` to use the compatibility shims to automatically convert iptables to nftables)

The main problem is that ``dockerd`` does a few things which make it incompatible (or to be more accurate, when you translate Docker's generated rules through the NFT translation utilities, it spits out something incompatible):

* It tries to update the ``ip`` table (IPv4 only) instead of the ``inet`` table (IPv4 + IPv6). The latter is the default table. Because it's updating the ``ip`` and not the ``inet`` table, it wants to add an input chain to that table. But an input chain already exists on ``inet`` table. So the effect of this is that the entire ``ip`` table gets ignored since you can't have an input chain on ``inet`` and on ``ip`` at the same time
* It tries to reference chains with capitalized lettering such as ``FORWARD`` instead of ``forward``. The default ``nftables`` identifiers for chains and other entities don't use upper case. 

In my testing, if these two things are fixed, Docker would work with ``nftables`` pretty seamlessly using ``iptables-nft``

Since this isn't supported currently, instead what we've had to do is create an ``/etc/nftables/docker.nft`` with the following process:

* Start Docker daemon and then run ``iptables-save > legacy-rules.txt``
* Convert to NFT: ``iptables-restore-translate -f legacy-rules.txt > new-rules.nft``
* Edit the rules and fix the issues mentioned above (``inet``, capitalization, etc.). You can remove some lines too, for example creation of ``inet`` tables, if that's already accomplished by the base rules.
* Move the rules to ``/etc/nftables/docker.nft``.
* On RHEL, edit ``/etc/sysconfig/nftables.conf`` and add ``include "/etc/nftables/docker.nft"``
* Run ``systemctl restart nftables`` to install the new rules.
* Run ``nft list rulset`` and validate everything merged correctly. In particular, the ``DOCKER*`` chains should be within the ``inet`` table, and the ``forward`` chain (not the ``FORWARD`` chain) should have updates.

The downside of this approach is ``dockerd`` is no longer responsible for making these updates automatically. So if the daemon's firewall design changes in some way, users now need to retrofit it to the above process

**Comment by FrankE-Aqura** [comment]
> nftables support is being worked on, however it is not a small change and cuts through just about every bit of the network stack.

Is there an ETA for a release that supports nftables?

**Comment by initiateit** [comment]
> nftables support is being worked on, however it is not a small change and cuts through just about every bit of the network stack.

Its not like we havent known nftables is replacing iptables for a long long time!

Still its nice to know thanks.

**Comment by FrankE-Aqura** [comment]
iptables and the iptables-nft package have been deprecated in RHEL 9: https://access.redhat.com/solutions/6739041
nftables is the default and recommended firewall in many distros now as well.

**Comment by TurnOffNOD** [comment]
The project netavark has nftables support now, and it claims:

> Netavark is a rust based network stack for containers. It is being designed to work with [Podman](https://github.com/containers/podman) but **is also applicable for other OCI container management applications**.

**Comment by pfeileon** [comment]
So what is the recommended way to deal with this issue until nftables is supported?

**Comment by hobeone** [comment]
In the short term if you use systemd you can link docker to the nftables service so it will restart when nftables is restarted.  Kind of a hack but it works.

On Debian this is what I added to my `/etc/systemd/system/docker.service.d/override.conf` file:

```
[Unit]
PartOf=nftables.service
```

I did this with my fail2ban service as well.

**Comment by pfeileon** [comment]
Thanks for your fast reply, but unfortunately I'm using docker-compose with podman.

**Comment by pfeileon** [comment]
So I got my setup (WSL2 Fedora 41 with podman and docker-compose) running again by changing the driver to iptables as shown on this [wiki-page (NetavarkNftablesDefault#How_To_Test)](https://fedoraproject.org/wiki/Changes/NetavarkNftablesDefault#How_To_Test). Simply replace the driver name with `iptables`.

Source: https://github.com/containers/podman/issues/24486#issuecomment-2461855407

**Comment by Talkless** [comment]
Here's usefull blog post by firewalld team:

**Comment by Talkless** [comment]
Here's useful blog post from firewalld:

https://firewalld.org/2024/04/strictly-filtering-docker-containers

It shows how to disable `iptables` in docker and then implement strict filtering in firewalld manually.

**Comment by bluikko** [comment]
> It shows how to disable `iptables` in docker and then implement strict filtering in firewalld manually.

How is it related to this issue?

**Comment by Talkless** [comment]
> How is it related to this issue?

If you want to use nftables, while docker uses iptables, you disable ~nftables~ **EDIT: iptables in Docker** and deal with nftables yourself via firewalld, and blog specifies Docker specifics what should be done.

**Comment by bluikko** [comment]
> > How is it related to this issue?
> 
> If you want to use nftables, while docker uses iptables, you disable nftables and deal with nftables yourself, and blog specifies Docker specifics what should be done.

This is a ridiculous suggestion. Do you understand that iptables is being deprecated everywhere and will be removed sooner rather than later? This means that Docker needs to support nftables sooner rather than later.

**Comment by jirislaby** [comment]
There are no `iptables` usually nowadays at all. Only nftables' iptables compatibility layer. That is not going away any time soon. But it's limited and users should be ported to nft (sooner than later).

**Comment by Talkless** [comment]
> This is a ridiculous suggestion. 

???

I am not suggesting to use iptables. firewalld supports nftables, you can use it to manage Docker and your own firewall rules.

**Comment by Talkless** [comment]
> you disable nftables 

Sorry, there was typo in my original post. You disable _iptables_ in Docker, and then use firewalld to configure _nftables_.

**Comment by rforberger** [comment]
Yeah, but what I actually want is, nft rules come up with docker restarting. 
I'll look into podman, if they support nftables natively.

**Comment by initiateit** [comment]
> > This is a ridiculous suggestion.
> 
> ???
> 
> I am not suggesting to use iptables. firewalld supports nftables, you can use it to manage Docker and your own firewall rules.

Nftables is the default firewall for Debian, ubuntu derivatives not to mention other distributions. Firewalld is redhat / fedora.

There is no reason for users to have to use workarounds, Docker need to get with the times and support Nftables out of the box. Workarounds eventually cause breakage.

**Comment by bluikko** [comment]
> > > This is a ridiculous suggestion.
> > 
> > 
> > ???
> > I am not suggesting to use iptables. firewalld supports nftables, you can use it to manage Docker and your own firewall rules.
> 
> Nftables is the default firewall for Debian, ubuntu derivatives not to mention other distributions. Firewalld is redhat / fedora.
> 
> There is no reason for users to have to use workarounds, Docker need to get with the times and support Nftables out of the box. Workarounds eventually cause breakage.

This is the reason this issue has almost a 100 +1.

**Comment by Talkless** [comment]
> Nftables is the default firewall for Debian, ubuntu derivatives not to mention other distributions. Firewalld is redhat / fedora.

It's nftables everywhere. Firewalld, UFW, the new foomuuri - they all just generate nftables rules. There's no "firewalld", "UFW" in the Linux kernel.

>  Docker need to get with the times and support Nftables

Of course, I do not downplay the enormous need to support nftables. Just noted that there's a workaround for the needed.

**Comment by Talkless** [comment]
> Yeah, but what I actually want is, nft rules come up with docker restarting.

ChatGPT suggest:

```
sudo systemctl edit docker.service
```

```
[Service]
ExecStartPre=/usr/local/bin/docker-firewall-pre.sh
ExecStopPost=/usr/local/bin/docker-firewall-post.sh
```
And these two scripts:

```
#!/bin/bash
firewall-cmd --zone=dockerzone --add-source=172.17.0.0/16 --permanent
firewall-cmd --reload

```

```
#!/bin/bash
firewall-cmd --zone=dockerzone --remove-source=172.17.0.0/16 --permanent
firewall-cmd --reload
```

Firewalld zone is "inactive" once it does not have "interface" or "source" assigned.

Since zone will have no interface/source, these "policies" (FORWARD/OUTPUT rules) generated by firewalld as in that blog post, should not match anything I guess.

**Comment by pdf** [comment]
I realise the irony of adding another comment to say this, but if you could stop spamming all subscribers that are tracking this issue for completion of the reported issue (first-party nftables support in Docker) with irrelevancies and workarounds (especially LLM-generated slop that does not assist at all), that would be greatly appreciated.

**Comment by kid1412621** [comment]
> nftables support is being worked on, however it is not a small change and cuts through just about every bit of the network stack.

hi, just wondering any update? since 1y passed

**Comment by ghost** [comment]
there is a lot of problem related to the iptable_nft
for example
https://stackoverflow.com/questions/57710177/tproxy-compatibility-with-docker

also if you deploy docker on routers like openwrt that uses nftable by default, and using "ip6tables":true in daemon.json, ipv6 forward will stop working on the LAN network, even after stop nerdctl and run ip6tables -F.

the router has to be rebooted without docker to make ipv6 of LAN work again

**Comment by lifeofguenter** [comment]
Would it not be sufficient as workaround for dockerd to re-run iptables on reload?

Then I just need to reload dockerd after applying nftables.conf.

I do the same with fail2ban.

**Comment by pdf** [comment]
@lifeofguenter this issue is about native support.

It is possible to play games with systemd units to restart everything that touches firewall rules, but this is far from ideal, and currently docker upgrades fail in the presence of nftables by default, even with compatibility tables in place.

**Comment by initiateit** [comment]
I have moved on from docker and Incus [https://linuxcontainers.org/incus/](url) is what I now use. It has the ability to use OCI so there is no issue running existing docker images. 

It has full nftables support. Docker have no interest in fixing this.

**Comment by bluikko** [comment]
People need to stop polluting this thread with half-baked workarounds.

**Comment by initiateit** [comment]
@bluikko it aint going to get fixed. There is no business case. They are happy just to leave a wrapper for iptables.

**Comment by akerouanton** [comment]
Our initial plan was to include nftables support in v29, but it's not going to make it as the first RC is scheduled for July 10 (although this date might slip). The implementation itself is in good shape, but it needs thorough review before it can be merged and released. We've an Epic open here to track the work left: https://github.com/moby/moby/issues/49634.

Firewalld will get its own implementation as we're currently relying on [firewalld.direct(5)](https://firewalld.org/documentation/man-pages/firewalld.direct.html) which is deprecated. This is likely going to be implemented after the initial nftables release though.

At the moment, the best (and only) way to go is `iptables-nft`. There's no other workaround one can use that gives the same level of network isolation.

Please refrain from commenting here, or on the Epic, if you don't have anything to add. The signal:noise ratio has been pretty low, and that would warrant locking this issue if it continues.

**Comment by qhaas** [comment]
[RHEL 9 is issuing these warnings](https://access.redhat.com/solutions/7123415) about modules from the `iptables-nft` package required by `docker-ce`
```
Warning: Unmaintained driver is detected: nft_compat
Warning: Unmaintained driver is detected: ip_set
```

The RHEL documentation claims they are deprecated, yet the warning implies unmaintained (two different things), but if they are no longer getting security updates, this will eventually start dinging security scans.

**Comment by TurnOffNOD** [comment]
> [RHEL 9 is issuing these warnings](https://access.redhat.com/solutions/7123415) about modules from the `iptables-nft` package required by `docker-ce`
> 
> ```
> Warning: Unmaintained driver is detected: nft_compat
> Warning: Unmaintained driver is detected: ip_set
> ```
> 
> The RHEL documentation claims they are deprecated, yet the warning implies unmaintained (two different things), but if they are no longer getting security updates, this will eventually start dinging security scans.

Maybe because it is deprecated, so the developers lose interest on it and move away to other open-source projects, and then it becomes unmaintained? 

(Just a joke)

**Comment by rwahowa** [comment]
Greatest news ever, docker now supports nftables (experimental for now) . I took it for a spin and it will be my way forward. Here is the video of me [ trying docker with nftables for the first time ](https://youtu.be/WTmld95BxHU )

**Comment by encbladexp** [comment]
> Greatest news ever, docker now supports nftables (experimental for now) . I took it for a spin and it will be my way forward. Here is the video of me [ trying docker with nftables for the first time ](https://youtu.be/WTmld95BxHU)

We should watch a 37 minutes video? SRSLY?

Just [this link](https://docs.docker.com/engine/network/firewall-nftables/) provides the same information, including the ability to use copy & paste. 🪄 

Service note: Docker v29 will have experimental Support. 🧑‍🔬

**Comment by rwahowa** [comment]
Added timelines to enable you skip about, based on what you're looking for.

....................................
Check out Bizanosa.com for more information about our Linux server support
services and pricing .

**Comment by akerouanton** [comment]
@rwahowa As maintainers we appreciate community members discussing and promoting new Engine features, but Github issues aren't the right place for self-promotion. I already marked your previous comment as spam, and I'm going to mark your last one as spam too.

**Comment by stefanpejcic** [comment]
this was just added in [29.0.0](https://docs.docker.com/engine/release-notes/29/#2900) 🎉

> Experimental support for nftables can now be enabled by setting Docker daemon's firewall-backend option to nftables. For more information, see [Docker Engine docs](https://docs.docker.com/engine/network/firewall-nftables/).

**Comment by rwahowa** [comment]
> [@rwahowa](https://github.com/rwahowa) As maintainers we appreciate community members discussing and promoting new Engine features, but Github issues aren't the right place for self-promotion. I already marked your previous comment as spam, and I'm going to mark your last one as spam too.

Noted. Understood

**Comment by robmry** [comment]
I'll mark this as closed - nftables is opt-in now, apart from Swarm which should also be supported soon (and is tracked separately) ...
- https://github.com/moby/moby/issues/49634
